agent/download.go

```diff
@@ -73,38 +74,43 @@ func (d Download) Start(ctx context.Context) error {
 	})
 }
 
-func getTargetPath(path string, destination string) string {
+func targetPath(ctx context.Context, dlPath, destPath string) string {
+	dlPath = filepath.Clean(dlPath)
+
 	// If we're downloading a file with a path of "pkg/foo.txt" to a folder
 	// called "pkg", we should merge the two paths together. So, instead of it
 	// downloading to: destination/pkg/pkg/foo.txt, it will just download to
 	// destination/pkg/foo.txt
-	destinationPaths := strings.Split(destination, string(os.PathSeparator))
-	downloadPaths := strings.Split(path, string(os.PathSeparator))
-
-	for i := 0; i < len(downloadPaths); i += 100 {
-		// If the last part of the destination path matches
-		// this path in the download, then cut it out.
-		lastIndex := len(destinationPaths) - 1
-
-		// Break if we've gone too far.
-		if lastIndex == -1 {
-			break
-		}
+	destPathComponents := strings.Split(destPath, string(os.PathSeparator))
+	dlPathComponents := strings.Split(dlPath, string(os.PathSeparator))
+
+	// If the last component of the destination path matches the first component
+	// of the download path, then trim the last component of the destination.
+	lastIndex := len(destPathComponents) - 1
+	lastDestComponent := destPathComponents[lastIndex]
+	if lastDestComponent == dlPathComponents[0] {
```

Comment:
not relevant to this PR, but this isn't mentioned anywhere in the doc: https://buildkite.com/docs/agent/v3/cli-artifact#downloading-artifacts-artifact-download-examples

😮